crates/bevy_input/src/touch.rs

```diff
@@ -0,0 +1,99 @@
+use bevy_app::{EventReader, Events};
+use bevy_ecs::{Local, Res, ResMut};
+use bevy_math::Vec2;
+use std::collections::{HashMap, HashSet};
+
+/// A touch input event
+#[derive(Debug, Clone)]
+pub struct TouchInput {
+    pub phase: TouchPhase,
+    pub position: Vec2,
+    ///
+    /// ## Platform-specific
+    ///
+    /// Unique identifier of a finger.
+    pub id: u64,
+}
+
+/// Describes touch-screen input state.
+#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+pub enum TouchPhase {
+    Started,
+    Moved,
+    Ended,
+    Cancelled,
+}
+
+#[derive(Default)]
+pub struct TouchSystemState {
+    touch_event_reader: EventReader<TouchInput>,
+}
+
+#[derive(Debug, Clone)]
+pub struct ActiveTouch {
+    start_position: Vec2,
+    previous_position: Vec2,
+    current_position: Vec2,
+}
+
+impl ActiveTouch {
+    pub fn delta(&self) -> Vec2 {
+        self.current_position - self.previous_position
+    }
+
+    pub fn distance(&self) -> Vec2 {
+        self.current_position - self.start_position
+    }
+}
+
+#[derive(Default)]
+pub struct TouchInputState {
```

Comment:
I'm thinking that we should make this the "go to" high level interface for touch interaction (much like the `Input<T>` abstractions). This would reduce the need to directly consume events.

Some thoughts on changes to make here:

* Rename `TouchInputState`, to the more friendly `Touches` 
* Rename `ActiveTouch` to just `Touch`
* Add the "touch id" to ActiveTouch
* Add an iter() method to Touches, which iterates all Touches
* Make all Touches fields private + add relevant methods to expose functionality

That would result in user code looking something like this:

```rust

fn touch_system(touches: Res<Touches>) {
    for touch in touches.iter() {
        println!(
            "active touch: {} {} {} {}",
            touch.id, touch.position, touch.previous_position, touch.start_position
        );

        if touches.just_pressed(touch.id) {
            println!(
                "just pressed touch with id: {:?}, at: {:?}",
                touch.id, touch.position
            );
        }

        if touches.just_released(touch.id) {
            println!(
                "just released touch with id: {:?}, at: {:?}",
                touch.id, touch.position
            );
        }

        if touches.just_cancelled(touch.id) {
            println!("cancelled touch with id: {:?}", event.id);
        }
    }
}
```

Reply:
Ex:

```rust

#[derive(Default)]
pub struct Touches {
    active_touches: HashMap<u64, Touch>,
    just_pressed: HashSet<u64>,
    just_released: HashSet<u64>,
    just_cancelled: HashSet<u64>,
}

impl Touches {
    pub fn iter(&self) -> impl Iterator<Item=&ActiveTouch> + '_ {
        self.active_touches.values()
    }

    pub fn just_pressed(&self, id: u64) -> bool {
        self.just_pressed.contains(&id)
    }

    pub fn iter_just_pressed(&self) -> impl Iterator<Item=&ActiveTouch> + '_ {
        self.just_pressed.iter().map(move |id| self.active_touches.get(id).unwrap())
    }
}
```